### DISCUSSION TITLE
Can MOOSE make several discrete values into a new variable?

### DISCUSSION THREAD

**Discussion by Ethan-xj** [idaholab/moose#22128] [Q&A General] [answered]
I use the [NodalValueSampler](https://mooseframework.inl.gov/source/vectorpostprocessors/NodalValueSampler.html) and get the values of the specific variable at each node in a specific 1D-block(in the form of array). Now I want to make these values into a new variable and at the same time the nodal values of the new variable are the elements of the array I get. How can achieve it?

**Comment by GiudGiud** [comment, marked as answer]
not sure how to do it cleanly. 
seems parsedAux wont let you use a postprocessor

you could put it in a VectorPostprocessor and then use this
https://mooseframework.inl.gov/docs/site/source/auxkernels/VectorPostprocessorVisualizationAux.html

but there's got to be another way

**Reply by Ethan-xj** [reply]
Thanks a lot for your apply! I will have a try.And If I find some other way to achieve it I will tell you.

**Reply by Ethan-xj** [reply]
@GiudGiud Hello! I notice that the [VectorPostprocessorVisualizationAux](https://mooseframework.inl.gov/docs/site/source/auxkernels/VectorPostprocessorVisualizationAux.html) reads values from a VectorPostprocessor that is producing vectors that are **'number of processors' * in length**. So VectorPostprocessorVisualizationAux is used to put **the value for each processor** into auxiliary field. But to be honest, I'm not so familiar with the info about the processor in MOOSE. There is only one processor in my input file and what I want is to put the vector which I get from the VectorPostprocessor(NodalValueSampler) into auxiliary field.
Should I change the source code of the VectorPostprocessorVisualizationAux to achieve my goal?

**Reply by Ethan-xj** [reply]
@GiudGiud I've used the VectorPostprocessorVisualizationAux and I wrote an input file. But it shows an error:
```
*** **ERROR** ***
**The following error occurred in the object "AuxVELADJ", of type "VectorPostprocessorVisualizationAux".
Error in AuxKernel AuxVELADJ. Vector u in VectorPostprocessor VELADJ_VEC does not contain num_procs number of entries.  num_procs: 1 num_entries: 0**
Here is my input file:
[GlobalParams]
  gravity = -9.8
  transient_term = false
[]

[Mesh]
 file = 'twopipe.e'
[]

[Variables]
  [./p]
    order = SECOND
    family = LAGRANGE
    block = '1 2'
  [../]
  [./u]
    order = SECOND
    family = LAGRANGE
    block = '1 2'
    [./InitialCondition]
      type = ConstantIC
      value = 0.5
    [../]
  [../]
  [./T]
    order = SECOND
    family = LAGRANGE
    block = '1 2'
    [./InitialCondition]
      type = ConstantIC
      value = 628
    [../]
  [../]
  [./w]
    order = SECOND
    family = LAGRANGE
    block = '1'
  [../]
[]

[AuxVariables]
  [./VELADJ]
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./PRESSADJ]
    family = MONOMIAL
    order = CONSTANT
  [../]
  [./RHOADJ]
    family = MONOMIAL
    order = CONSTANT
  [../]
[]
[AuxKernels]
 [./AuxVELADJ]
   type = VectorPostprocessorVisualizationAux
   vpp = VELADJ_VEC
   vector_name = u
   variable = VELADJ
   execute_on = 'LINEAR NONLINEAR'
 [../]
 [./AuxPRESSADJ]
   type = VectorPostprocessorVisualizationAux
   vpp = PRESSADJ_VEC
   vector_name = p
   variable = PRESSADJ
   execute_on = 'LINEAR NONLINEAR'
 [../]
 [./AuxRHOADJ]
   type = VectorPostprocessorVisualizationAux
   vpp = RHOADJ_VEC
   vector_name = 'rho'
   variable = RHOADJ
   execute_on = 'LINEAR NONLINEAR'
 [../]
[]

[Kernels]
  [./MASS_EQUATION]
    type = mass
    variable = p
    vel = u
    temp = T
    press = p
    block = '1 2'
  [../]
  [./Momentum_EQUATION]
    type = momentum
    variable = u
    press = p
    temp = T
    vel = u
    block = '1 2'
   [../]
  [./Energy_EQUATION]
    type = energy
    variable = T
    vel = u
    press = p
    temp = T
    value = 1.0
    block = '1 2'
    function = test
  [../]
  [./LateralMomentum_EQUATION]
    type = lateralmomentum
    variable = w
    lvel = w
    vel = u
    temp = T
    press = p
    veladj = VELADJ
    pressadj = PRESSADJ
    rhoadj = RHOADJ
    block = '1'
    lk = 1
    sk = 0.5
    kg = 0.5
  [../]
[]

[Functions]
  [./test]
    type = ParsedFunction
    value = 1.0e8*sin(pi*x)
  [../]
[../]

[BCs]
  [./outlet_p]
    type =  DirichletBC
    boundary = '16 26'
    variable = p
    value = 2.0e5
  [../]
  [./inlet_u1]
    type =  DirichletBC
    boundary = '11'
    variable = u
    value = 1
  [../]
  [./inlet_u2]
    type =  DirichletBC
    boundary = '26'
    variable = u
    value = 1.5
  [../]
  [./inlet_T1]
    type = DirichletBC
    variable = T
    boundary = '11'
    value = 500
  [../]
  [./inlet_T2]
    type = DirichletBC
    variable = T
    boundary = '21'
    value = 600
  [../]
[]

[Materials]
  [./EM]
    type = submaterial
    is_sodium = true
    constant_property = false
    constant_f_h_property = true
    f_coefficient = 0.01
    diameter = 0.02
    temp = T
    block = '1 2'
  [../]
[]

[Preconditioning]
  [./FD]
    type = SMP
    full = true
  [../]
[]

[VectorPostprocessors]
  [./VELADJ_VEC]
    type = NodalValueSampler
    variable = u
    block = '2'
    sort_by = x
    execute_on = 'LINEAR NONLINEAR'
  [../]
  [./PRESSADJ_VEC]
    type = NodalValueSampler
    variable = p
    block = '2'
    sort_by = x
    execute_on = 'LINEAR NONLINEAR'
  [../]
  [./RHOADJ_VEC]
    type = Sampler1DReal
    block = '2'
    property = 'rho'
    sort_by = x
    execute_on = 'LINEAR NONLINEAR'
  [../]
[]

[Executioner]
  type =  Steady
  solve_type = PJFNK

  petsc_options = '-snes_converged_reason -ksp_converged_reason'
  petsc_options_iname = '-pc_type -pc_factor_shift_type -pc_factor_mat_solver_package'
  petsc_options_value = 'lu    NONZERO      superlu_dist'

  nl_rel_tol = 1e-8
  nl_abs_tol = 1e-12
  nl_max_its = 100
  l_tol = 1e-6
  l_max_its = 50
  line_search = 'none'
[]

[Outputs]
  exodus = true
  csv = true
```
I don't know why the num_entries is 0.

**Reply by GiudGiud** [reply]
The NodalValueSampler isnt going to have 1 entry. It ll have a lot. Not sure why it s reporting 0.

I would use a postprocessor to get the value you wasnt, then convert it to a VPP using the VectorOfPostprocessors vpp.

Anyway, I think we may be astray here.


You're just trying to copy a variable into another variable right?
Both the ParsedAux and SelfAux can do that

**Reply by Ethan-xj** [reply]
Actually the "variable" and the "another variable" are from different blocks. The [Variable] and [Kernel] block in MOOSE are all block-restrictable. So I meet this problem when I try to express "variable in another block" in the current block. And why I have to do that is because the equation which I want to solve is semiempirical. The equation is independent of grid.
All in all, thank you a lot for your patience! I will try SelfAux first.

**Reply by GiudGiud** [reply]
ah ok. 
Can they be split in a different calculation using the MultiApp system? 
This is typically how we do this.

SelfAux would only work if they were defined on the same blocks
